x-pack/plugins/security/server/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,271 @@
+
+    const [, kerberosTicket] = (request.headers.authorization as string).split(/\s+/);
+
+    // First attempt to exchange SPNEGO token for an access token.
+    let tokens: { access_token: string; refresh_token: string };
+    try {
+      tokens = await this.options.client.callAsInternalUser('shield.getAccessToken', {
+        body: { grant_type: '_kerberos', kerberos_ticket: kerberosTicket },
+      });
+    } catch (err) {
+      this.logger.debug(`Failed to exchange SPNEGO token for an access token: ${err.message}`);
+      return AuthenticationResult.failed(err);
+    }
+
+    this.logger.debug('Get token API request to Elasticsearch successful');
+
+    try {
+      // Then attempt to query for the user details using the new token
+      const authHeaders = { authorization: `Bearer ${tokens.access_token}` };
+      const user = await this.getUser(request, authHeaders);
+
+      this.logger.debug('User has been authenticated with new access token');
+      return AuthenticationResult.succeeded(user, {
+        authHeaders,
+        state: { accessToken: tokens.access_token, refreshToken: tokens.refresh_token },
+      });
+    } catch (err) {
+      this.logger.debug(`Failed to authenticate request via access token: ${err.message}`);
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to authenticate request with `Bearer ***` Authorization header by passing it to the Elasticsearch backend.
+   * @param request Request instance.
+   */
+  private async authenticateWithBearerScheme(request: KibanaRequest) {
+    this.logger.debug('Trying to authenticate request using "Bearer" authentication scheme.');
+
+    try {
+      const user = await this.getUser(request);
+
+      this.logger.debug('Request has been authenticated using "Bearer" authentication scheme.');
+      return AuthenticationResult.succeeded(user);
+    } catch (err) {
+      this.logger.debug(
+        `Failed to authenticate request using "Bearer" authentication scheme: ${err.message}`
+      );
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to extract access token from state and adds it to the request before it's
+   * forwarded to Elasticsearch backend.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaState(request: KibanaRequest, { accessToken }: ProviderState) {
+    this.logger.debug('Trying to authenticate via state.');
+
+    if (!accessToken) {
+      this.logger.debug('Access token is not found in state.');
+      return AuthenticationResult.notHandled();
+    }
+
+    try {
+      const authHeaders = { authorization: `Bearer ${accessToken}` };
+      const user = await this.getUser(request, authHeaders);
+
+      this.logger.debug('Request has been authenticated via state.');
+      return AuthenticationResult.succeeded(user, { authHeaders });
```

Comment:
question: we're no longer mutating `request.headers` for this and all other token based auth providers which is great. have we ensured that we aren't breaking reporting by doing so?

Reply:
I'll double check before I ask for review, but I believe we should be fine since we maintain BWC behavior with this https://github.com/elastic/kibana/blob/master/src/core/server/http/http_server.ts#L238